File: src/main/scala/im/tox/toktok/app/message_activity/MessageAdapter.scala
package im.tox.toktok.app.message_activity

import android.support.v7.widget.RecyclerView
import android.util.SparseBooleanArray
import android.view.{ LayoutInflater, View, ViewGroup }
import android.widget.{ RelativeLayout, TextView }
import com.typesafe.scalalogging.Logger
import de.hdodenhof.circleimageview.CircleImageView
import im.tox.toktok.TypedResource._
import im.tox.toktok.app.domain.{ Message, MessageType }
import im.tox.toktok.app.message_activity.MessageAdapter.inflate
import im.tox.toktok.{ TR, TypedLayout }
import org.slf4j.LoggerFactory

import scala.collection.mutable.ListBuffer

object MessageAdapter {
  private def inflate[A <: View](layout: TypedLayout[A], viewGroup: ViewGroup): A = {
    LayoutInflater.from(viewGroup.getContext).inflate(layout, viewGroup, false)
  }
}

final class MessageAdapter(
    messageClick: MessageClick,
    messageActionMode: MessageActionMode
) extends RecyclerView.Adapter[MessageViewHolder] {

  val logger = Logger(LoggerFactory.getLogger(getClass))

  private val messages = new ListBuffer[Message]
  private val selectedItems = new SparseBooleanArray()
  private var actionModeActive = false

  override def getItemViewType(position: Int): Int = {
    messages(position).msgType.viewType
  }

  override def onCreateViewHolder(viewGroup: ViewGroup, viewType: Int): MessageViewHolder = {
    viewType match {
      case MessageType.Delivered.viewType =>
        val itemView = inflate(TR.layout.message_item_user_simple, viewGroup)
        new MessageViewHolderDetailed(itemView, messageActionMode)
      case MessageType.Received.viewType =>
        val itemView = inflate(TR.layout.message_item_friend_simple, viewGroup)
        new MessageViewHolderDetailed(itemView, messageActionMode)
      case MessageType.Action.viewType =>
        val itemView = inflate(TR.layout.message_item_action, viewGroup)
        new MessageViewHolder(itemView, messageActionMode, itemView.findView(TR.message_item_action))
    }
  }

  def toggleSelection(i: Int): Unit = {
    if (selectedItems.get(i, false)) {
      selectedItems.delete(i)
    } else {
      selectedItems.put(i, true)
    }
    notifyItemChanged(i)
  }

  def getSelectedItemCount: Int = {
    selectedItems.size()
  }

  def isSelected(position: Int): Boolean = {
    selectedItems.get(position, false)
  }

  override def onBindViewHolder(viewHolder: MessageViewHolder, position: Int): Unit = {
    val message = messages(position)

    viewHolder.mUserText.setText(message.msgContent)
    viewHolder.mUserImg.setImageResource(message.imageSrc)

    if (actionModeActive && !isSelected(position)) {
      viewHolder.base.setAlpha(0.5f)
    } else {
      viewHolder.base.setAlpha(1)
    }

    getItemViewType(position) match {
      case MessageType.Delivered.viewType =>
        val view = viewHolder.asInstanceOf[MessageViewHolderDetailed]
        view.mUserDetails.setText(message.msgDetails)

      case MessageType.Received.viewType =>
        viewHolder.mUserImg.setOnClickListener(new View.OnClickListener {
          override def onClick(v: View): Unit = {
            messageClick.onImgClick()
          }
        })

        val view = viewHolder.asInstanceOf[MessageViewHolderDetailed]
        view.mUserDetails.setText(message.msgDetails)

      case MessageType.Action.viewType =>
        viewHolder.mUserImg.setOnClickListener(new View.OnClickListener {
          override def onClick(v: View): Unit = {
            messageClick.onImgClick()
          }
        })
    }
  }

  def getItemCount: Int = {
    messages.length
  }

  def addItem(newMsg: Message): Unit = {
    messages.insert(0, newMsg)
    notifyItemInserted(0)
  }

  def setActionModeActive(state: Boolean): Unit = {
    actionModeActive = state
    logger.debug(actionModeActive.toString)
  }

  def clearSelections(): Unit = {
    selectedItems.clear()
    notifyDataSetChanged()
  }

  def deleteSelected(): Unit = {
    for (a <- (messages.size - 1) to 0 by -1) {
      if (selectedItems.get(a)) {
        messages.remove(a)
        notifyItemRemoved(a)
      }
    }

    selectedItems.clear()
  }

}

sealed class MessageViewHolder(
    itemView: RelativeLayout,
    messageActionMode: MessageActionMode,
    final val base: View
) extends RecyclerView.ViewHolder(itemView)
  with View.OnLongClickListener
  with View.OnClickListener {

  final val mUserText: TextView = itemView.findView(TR.message_item_text)
  final val mUserImg: CircleImageView = itemView.findView(TR.message_item_img)

  itemView.setOnLongClickListener(this)
  itemView.setOnClickListener(this)

  def onLongClick(v: View): Boolean = {
    messageActionMode.onLongClick(getLayoutPosition)
  }

  def onClick(v: View): Unit = {
    messageActionMode.onClick(getLayoutPosition)
  }

}

final class MessageViewHolderDetailed(
    itemView: RelativeLayout,
    messageActionMode: MessageActionMode
) extends MessageViewHolder(
  itemView,
  messageActionMode,
  itemView.findView(TR.message_item_base)
) {
  val mUserDetails: TextView = itemView.findView(TR.message_item_details)
}

trait MessageClick {
  def onImgClick(): Unit
}

trait MessageActionMode {
  def onLongClick(i: Int): Boolean
  def onClick(i: Int): Unit
}

File: sources/app/src/main/java/im/tox/toktok/app/message_activity/MessageAdapter.java
package im.tox.toktok.app.message_activity;

import android.util.SparseBooleanArray;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.RelativeLayout;
import android.widget.TextView;

import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.ArrayList;
import java.util.List;

import de.hdodenhof.circleimageview.CircleImageView;
import im.tox.toktok.R;
import im.tox.toktok.app.domain.Message;

final class MessageAdapter extends RecyclerView.Adapter<MessageViewHolder> {

    private static final Logger logger = LoggerFactory.getLogger(MessageAdapter.class);

    private final MessageClick messageClick;
    private final MessageActionMode messageActionMode;
    private final List<Message> messages = new ArrayList<>();
    private final SparseBooleanArray selectedItems = new SparseBooleanArray();
    private boolean actionModeActive = false;

    MessageAdapter(
            MessageClick messageClick,
            MessageActionMode messageActionMode
    ) {
        this.messageClick = messageClick;
        this.messageActionMode = messageActionMode;
    }

    @NonNull
    private static RelativeLayout inflate(int layout, @NonNull ViewGroup viewGroup) {
        return (RelativeLayout) LayoutInflater.from(viewGroup.getContext())
                .inflate(layout, viewGroup, false);
    }

    @Override
    public int getItemViewType(int position) {
        return messages.get(position).msgType.viewType;
    }

    @Override
    public MessageViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        RelativeLayout itemView;
        switch (viewType) {
            case 1: //MessageType.Delivered:
                itemView = inflate(R.layout.message_item_user_simple, viewGroup);
                return new MessageViewHolderDetailed(itemView, messageActionMode);
            case 2: //MessageType.Received.viewType =>
                itemView = inflate(R.layout.message_item_friend_simple, viewGroup);
                return new MessageViewHolderDetailed(itemView, messageActionMode);
            case 3: //MessageType.Action.viewType =>
                itemView = inflate(R.layout.message_item_action, viewGroup);
                return new MessageViewHolder(itemView, messageActionMode, itemView.findViewById(R.id.message_item_action));
            default:
                throw new RuntimeException("OMG");
        }
    }

    void toggleSelection(int i) {
        if (selectedItems.get(i, false)) {
            selectedItems.delete(i);
        } else {
            selectedItems.put(i, true);
        }
        notifyItemChanged(i);
    }

    int getSelectedItemCount() {
        return selectedItems.size();
    }

    private boolean isSelected(int position) {
        return selectedItems.get(position, false);
    }

    @Override
    public void onBindViewHolder(@NonNull MessageViewHolder viewHolder, int position) {
        Message message = messages.get(position);

        viewHolder.mUserText.setText(message.msgContent);
        viewHolder.mUserImg.setImageResource(message.imageSrc);

        if (actionModeActive && !isSelected(position)) {
            viewHolder.base.setAlpha(0.5f);
        } else {
            viewHolder.base.setAlpha(1);
        }
        MessageViewHolderDetailed view;
        switch (getItemViewType(position)) {
            case 1://MessageType.Delivered.viewType =>
                view = (MessageViewHolderDetailed) viewHolder;
                view.mUserDetails.setText(message.msgDetails);

            case 2://MessageType.Received.viewType =>
                viewHolder.mUserImg.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        messageClick.onImgClick();
                    }
                });

                view = (MessageViewHolderDetailed) viewHolder;
                view.mUserDetails.setText(message.msgDetails);

            case 3://MessageType.Action.viewType =>
                viewHolder.mUserImg.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        messageClick.onImgClick();
                    }
                });
        }
    }

    @Override
    public int getItemCount() {
        return messages.size();
    }

    void addItem(Message newMsg) {
        messages.add(0, newMsg);
        notifyItemInserted(0);
    }

    void setActionModeActive(boolean state) {
        actionModeActive = state;
        logger.debug(String.valueOf(actionModeActive));
    }

    void clearSelections() {
        selectedItems.clear();
        notifyDataSetChanged();
    }

    void deleteSelected() {
        for (int a = messages.size() - 1; a >= 0; a--) {
            if (selectedItems.get(a)) {
                messages.remove(a);
                notifyItemRemoved(a);
            }
        }

        selectedItems.clear();
    }

}

class MessageViewHolder extends RecyclerView.ViewHolder
        implements View.OnLongClickListener
        , View.OnClickListener {

    private final MessageActionMode messageActionMode;

    final TextView mUserText;
    final CircleImageView mUserImg;
    final View base;

    MessageViewHolder(
            @NonNull RelativeLayout itemView,
            MessageActionMode messageActionMode,
            View base
    ) {
        super(itemView);
        mUserText = itemView.findViewById(R.id.message_item_text);
        mUserImg = itemView.findViewById(R.id.message_item_img);
        this.messageActionMode = messageActionMode;
        this.base = base;
        itemView.setOnLongClickListener(this);
        itemView.setOnClickListener(this);
    }

    @Override
    public boolean onLongClick(View v) {
        return messageActionMode.onLongClick(getLayoutPosition());
    }

    @Override
    public void onClick(View v) {
        messageActionMode.onClick(getLayoutPosition());
    }

}

final class MessageViewHolderDetailed extends MessageViewHolder {

    final TextView mUserDetails;

    MessageViewHolderDetailed(
            @NonNull RelativeLayout itemView,
            MessageActionMode messageActionMode
    ) {
        super(
                itemView,
                messageActionMode,
                itemView.findViewById(R.id.message_item_base)
        );
        mUserDetails = itemView.findViewById(R.id.message_item_details);
    }
}

interface MessageClick {
    void onImgClick();
}

interface MessageActionMode {
    boolean onLongClick(int i);

    void onClick(int i);
}
